src/calliope/postprocess/orchestrator.py

```diff
@@ -0,0 +1,175 @@
+# Copyright (C) since 2013 Calliope contributors listed in AUTHORS.
+# Licensed under the Apache 2.0 License (see LICENSE file).
+"""Helper registry definition for extensible model postprocessing."""
+
+from __future__ import annotations
+
+import logging
+import threading
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from typing import Final, TypeVar
+
+import numpy as np
+import xarray as xr
+from pydantic import BaseModel, ConfigDict
+
+from calliope.exceptions import ModelError
+from calliope.schemas.config_schema import CalliopeConfig
+from calliope.schemas.general import AttrStr, NonEmptyUniqueList
+from calliope.util.tools import listify
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, slots=True)
+class PostprocessContext:
+    """Small runtime context handler for postprocessing functions."""
+
+    inputs: xr.Dataset
+    results: xr.Dataset
+    config: CalliopeConfig
+
+
+PostprocessFunction = Callable[[PostprocessContext], xr.DataArray | None]
+F = TypeVar("F", bound=PostprocessFunction)
+
+
+class PostprocessSettings(BaseModel):
+    """Schema for postprocessing function settings."""
+
+    model_config = ConfigDict(frozen=True, extra="forbid")
+
+    name: AttrStr
+    order: int
+    base_math: NonEmptyUniqueList
+    enabled: bool
+
+
+class PostprocessRegistry:
+    """Postprocessing registry class."""
+
+    def __init__(self) -> None:
+        """Initialise with an empty registry."""
+        self._entries: dict[str, tuple[PostprocessSettings, PostprocessFunction]] = {}
+        self._lock = threading.Lock()
+
+    def register_function(
+        self,
+        function: PostprocessFunction,
+        *,
+        name: str,
+        base_math: Iterable[str],
```

Comment:
Does `base_math` _just_ look at `base`? Ideally it would look at the full list of `extra_math`/ `apply_math` and decide based on what the user has actually applied to their given use case.

Reply:
Although there will always be the danger of wanting to postprocess slightly differently between two instances applied math and them becoming conflicted (e.g. between `operate` and `plan`, but then you have _both_ defined in your applied math).